XADStuffItDESHandle.m

```diff
@@ -38,19 +38,33 @@ +(NSData *)keyForPasswordData:(NSData *)passworddata entryKey:(NSData *)entrykey
 	StuffItDESKeySchedule ks;
 
 	if(!mkey||[mkey length]!=8) [XADException raiseIllegalDataException];
-
-	uint8_t passblock[8]={0,0,0,0,0,0,0,0};
-	int length=[passworddata length];
-	if(length>8) length=8;
-	memcpy(passblock,[passworddata bytes],length);
-
+ 
 	// Calculate archive key and IV from password and mkey
-	uint8_t archivekey[8],archiveiv[8];
-
-	const uint8_t initialkey[8]={0x01,0x23,0x45,0x67,0x89,0xab,0xcd,0xef};
-	for(int i=0;i<8;i++) archivekey[i]=initialkey[i]^(passblock[i]&0x7f);
-	StuffItDESSetKey(initialkey,&ks);
-	StuffItDESCrypt(archivekey,&ks,YES);
+	uint8_t archiveiv[8];
+	uint8_t archivekey[8]={0x01,0x23,0x45,0x67,0x89,0xab,0xcd,0xef};
+	uint32_t x = 0;
+	uint32_t length=[passworddata length];
+  
+	StuffItDESSetKey(archivekey,&ks);
+	
+	do
+ 	{
+		 uint8_t passblock[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };
+
+		memcpy(passblock, [passworddata bytes] + (x << 3), 8);
```

Comment:
🤔 
this could lead to the reading out of bounds, in case if password is not multiple of 8

Reply:
Good catch! I fixed it in a commit... But now I have a bizarre issue where even on a default Ubuntu install, unar won't pass any password at all. Even a `printf("%d", length)` in this function yields 0. I could have sworn this was working a few weeks ago.